### PR TITLE
STI - clean test pass

### PR DESCRIPTION
**Pull request by GregSutcliffe**
This probably still needs work... but it's a start...

### CODE REVIEW

File: app/controllers/hosts_controller.rb [hunk 1]
```diff
@@ -137,7 +140,7 @@ def hostgroup_or_environment_selected
       @hostgroup   = Hostgroup.find(params[:hostgroup_id])     unless params[:hostgroup_id].empty?
       @host        = Host.find(params[:host_id])               if params[:host_id].to_i > 0
       if @environment or @hostgroup
-        @host ||= Host.new
+        @host ||= Host.new :type => params['host']['type']
```
Comment (ohadlevy) [inline app/controllers/hosts_controller.rb]: why do you need it? afair, new still accepts type?

File: app/controllers/hosts_controller.rb [hunk 2]
```diff
@@ -127,6 +129,7 @@ def compute_resource_selected
   end
 
   def hostgroup_or_environment_selected
+    logger.info params.inspect
```
Comment (ohadlevy) [inline app/controllers/hosts_controller.rb]: debug?

File: app/controllers/hosts_controller.rb [hunk 3]
```diff
@@ -494,6 +497,11 @@ def find_by_name
     deny_access and return unless User.current.admin? or Host.my_hosts.include?(@host)
   end
 
+  def set_base_type
+    params['host'] ||= {}
```
Comment (ohadlevy) [inline app/controllers/hosts_controller.rb]: I'm really not sure why you need this, is it because of tests?

File: app/helpers/hosts_helper.rb [hunk 4]
```diff
@@ -240,7 +240,8 @@ def host_title_actions(host, vm)
                                   :disabled => !Setting[:puppetrun],
                                   :title => "Trigger a puppetrun on a node; requires that puppet run is enabled"),
             link_to_if_authorized("All Puppet Classes", hash_for_storeconfig_klasses_host_path(:id => host).merge(:auth_action => :read),
-                                  :disabled => host.resources.count == 0,
+                                  :disabled => (host.resources.count == 0 rescue false),
```
Comment (ohadlevy) [inline app/helpers/hosts_helper.rb]: if you broke of puppet, then there is no longer resources and we should not show this btn

File: app/models/host.rb [hunk 5]
```diff
@@ -1,951 +1,11 @@
-require 'facts_importer'
+class Host < ActiveRecord::Base
```
Comment (ohadlevy) [inline app/models/host.rb]: this is confusing..

shouldnt this be host::base ?

**Comment by GregSutcliffe** [comment]
Clean test pass, no controller changes, index added to host.type. I think it's good to go! @ohadlevy over to you!

File: app/helpers/hosts_helper.rb [hunk 6]
```diff
@@ -238,10 +238,7 @@ def host_title_actions(host, vm)
         button_group(
             link_to_if_authorized("Run puppet", hash_for_puppetrun_host_path(:id => host).merge(:auth_action => :edit),
                                   :disabled => !Setting[:puppetrun],
-                                  :title => "Trigger a puppetrun on a node; requires that puppet run is enabled"),
-            link_to_if_authorized("All Puppet Classes", hash_for_storeconfig_klasses_host_path(:id => host).merge(:auth_action => :read),
-                                  :disabled => host.resources.count == 0,
-                                  :title => "Show all host puppet classes, requires storeconfigs")
+                                  :title => "Trigger a puppetrun on a node; requires that puppet run is enabled")
```
Comment (ohadlevy) [inline app/helpers/hosts_helper.rb:241]: it should be clearly stated that this commit also removes the store config integration between foreman & puppet?

File: app/models/domain.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ class Domain < ActiveRecord::Base
   include Authorization
   include Taxonomix
 
-  has_many :hosts
+  has_many :hosts, :class_name => 'Host::Managed'
```
Comment (ohadlevy) [inline app/models/domain.rb]: I wonder if it makes any sense to make it a mixin - so later on when we change it we don't need to do it again

File: app/models/host.rb [hunk 8]
```diff
@@ -1,951 +1,27 @@
...
-        values   = facts.values
-      when Hash
-        certname = facts["certname"]
-        name     = facts["fqdn"]
-        values   = facts
-        return raise("invalid facts hash") unless name and values
+module Host
+
+  def self.method_missing(method, *args, &block)
+    super
```
Comment (ohadlevy) [inline app/models/host.rb]: does it make sense to call super here at all? we already matched a method_missing ? (e.g. if we got here, we already didn't find a method)?

File: app/models/puppetclass.rb [hunk 9]
```diff
@@ -6,7 +6,7 @@ class Puppetclass < ActiveRecord::Base
   has_many :hostgroup_classes, :dependent => :destroy
   has_many :hostgroups, :through => :hostgroup_classes
   has_many :host_classes, :dependent => :destroy
-  has_many :hosts, :through => :host_classes
+  has_many :hosts, :through => :host_classes, :foreign_key => :host_id
```
Comment (ohadlevy) [inline app/models/puppetclass.rb]: do we still need the foreign_key ?

File: app/models/token.rb [hunk 10]
```diff
@@ -12,7 +12,7 @@
 
 class Token < ActiveRecord::Base
   attr_accessible :value, :expires
-  belongs_to :host
+  belongs_to :host, :foreign_key => :host_id
```
Comment (ohadlevy) [inline app/models/token.rb:15]: stange, don't we need class_name? otherwise, how does it know the relationship?

File: app/models/user_fact.rb [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 class UserFact < ActiveRecord::Base
   belongs_to :user
-  belongs_to :fact_name, :class_name => "Puppet::Rails::FactName"
+  belongs_to :fact_name, :class_name => "FactName"
```
Comment (ohadlevy) [inline app/models/user_fact.rb]: I think you can drop the class_name here

File: app/helpers/hosts_helper.rb (hunk 6 above)
Comment (GregSutcliffe) [inline app/helpers/hosts_helper.rb:241]: Yes, I'll add that comment.

File: app/models/domain.rb (hunk 7 above)
Comment (GregSutcliffe) [inline app/models/domain.rb]: Sounds sensible. How difficult is that? I want to get this done today so I can enjoy my PTO, if possible :P

File: app/models/host.rb (hunk 8 above)
Comment (GregSutcliffe) [inline app/models/host.rb]: I guess not. Will fix.

File: app/models/token.rb (hunk 10 above)
Comment (GregSutcliffe) [inline app/models/token.rb:15]: Not sure, seems to work though:

```
irb(main):006:0> Host.last.token.to_s
=> "83c29bb2-33cf-401b-bd3c-c03c00562381"
```

File: app/models/user_fact.rb (hunk 11 above)
Comment (GregSutcliffe) [inline app/models/user_fact.rb]: Clean tests pass without that. Removing from the PR.

File: app/models/puppetclass.rb (hunk 9 above)
Comment (GregSutcliffe) [inline app/models/puppetclass.rb]: Seems fine without it:

```
irb(main):009:0> Host.first.puppetclasses.map{|p| p.try(:name)}
=> ["foreman", "puppet::server"]
```

Clean test run as well. Removing.

**Comment by GregSutcliffe** [comment]
@ohadlevy Mixin added (mostly copied from Taxonomix.rb). Seems to work for me, all tests pass and I can create/install libvirt VMs...

**Comment by GregSutcliffe** [comment]
Fact upload is broken, investigating....

**Comment by GregSutcliffe** [comment]
Fact upload fixed, puppet reports seem to work fine too.

**Comment by GregSutcliffe** [comment]
@ohadlevy latest commit is my attempt at this - I'm sure it can be cleaner. Seems to work though, I get one call to the DB for all the names, followed by the updates to FactValues.

**Comment by isratrade** [comment]
@ohadlevy, @GregSutcliffe, I pushed pr #430 which made come changes to Greg's branch.

**Pull request closed without merging**